### PR TITLE
Update impersonation role based on user attributes

### PR DESCRIPTION
**Pull request by cmurphy**
Problem 1: When the cluster role for the impersonation service account
was created by the role template binding controller, it only added rules
for two basic authenticated groups and optionally a group derived from
the role template binding, and ignored potential groups the user is
associated with in an external provider. This meant that a user logging
in through an external provider would be blocked from impersonating
their service account due to insufficient permissions.

Solution 1: This change updates the role template binding controller to
get the true known groups for the user from the attribute lister.

Problem 2: If the user's Extras attributes changed, such as what happens
when a local user becomes associated with an external provider, the
roles' rules were never be updated. This meant that even though the
groups from the provider were up to date in the role rules, the
userextras/username may be mismatched. It also means that if the user's
groups in the external provider changed at some point after the initial
role creation, they would become mismatched with the allowed groups in
the role.

Solution 2: In the impersonation package, check that the groups and user
extras in the UserInfo object match the rules for groups and extras in
the role, and update them if they do not match. When the setup is run
from the role template binding controller, the user data will come from
the role template, and when the setup is run from the proxy server, the
user data comes from the request token.

https://github.com/rancher/rancher/issues/33912

### CODE REVIEW

File: pkg/impersonation/impersonation.go [hunk 1]
```diff
@@ -182,10 +183,44 @@ func (i *Impersonator) createRole() (*rbacv1.ClusterRole, error) {
 			// in case cache isn't synced yet, use raw client
 			return i.clusterContext.RBAC.ClusterRoles("").Get(name, metav1.GetOptions{})
 		}
+		return r, nil
+	}
+	if !reflect.DeepEqual(rules, r.Rules) {
+		return i.updateRoleRules(r, rules[1], rules[2], rules[3])
```
Comment (ibuildthecloud) [inline pkg/impersonation/impersonation.go:189]: is there anything that guarentees that rules[2] and result[3] won't result in a panic?

File: pkg/impersonation/impersonation.go [hunk 2]
```diff
@@ -182,10 +183,44 @@ func (i *Impersonator) createRole() (*rbacv1.ClusterRole, error) {
...
+				role.Rules[i].ResourceNames = addToListIfNotPresent(rn, role.Rules[i].ResourceNames)
+			}
+		}
+		if role.Rules[i].Resources[0] == "userextras/username" {
+			for _, rn := range extrasUsernameRule.ResourceNames {
+				role.Rules[i].ResourceNames = addToListIfNotPresent(rn, role.Rules[i].ResourceNames)
+			}
+		}
+	}
+	return i.clusterContext.RBAC.ClusterRoles("").Update(role)
```
Comment (ibuildthecloud) [inline pkg/impersonation/impersonation.go:212]: Since this is additive and not making it equal, will that result in a situation in which each proxy API call you call update?  Basically first call `rules(A,B) != rules(C)` so you add C, then next call `rules(A, B, C) != rules(C)` so you don't modify the rule but still do an Update call?

File: pkg/impersonation/impersonation.go (hunk 1 above)
Comment (cmurphy) [reply pkg/impersonation/impersonation.go]: nope. fixed.

File: pkg/impersonation/impersonation.go (hunk 2 above)
Comment (cmurphy) [reply pkg/impersonation/impersonation.go]: fixed

File: pkg/impersonation/impersonation.go [hunk 3]
```diff
@@ -131,44 +132,44 @@ func (i *Impersonator) createNamespace() error {
...
+	}
+	if usernames, ok := extras["username"]; ok {
+		rules = append(rules, rbacv1.PolicyRule{
+			Verbs:         []string{"impersonate"},
+			APIGroups:     []string{"authentication.k8s.io"},
+			Resources:     []string{"userextras/username"},
+			ResourceNames: usernames,
+		})
+	}
+	role, err := i.clusterContext.RBAC.ClusterRoles("").Controller().Lister().Get("", name)
```
Comment (snasovich) [inline pkg/impersonation/impersonation.go:169]: Technically not part of the change itself, but it looks like `err` is not handled fully here - if `err` is not `IsNotFound` but is not `nil` we proceed as if there was no error.
Suggesting additional error handling to be added.

Comment (snasovich) [reply pkg/impersonation/impersonation.go:169]: Actually, disregard the "technically not part of the change itself" note. Previous code covered this error handling with `return r, err` at the end of func.

File: pkg/impersonation/impersonation.go [hunk 4]
```diff
@@ -34,17 +35,17 @@ func (i *Impersonator) SetUpImpersonation() (*corev1.ServiceAccount, error) {
 	if err != nil {
 		return nil, err
 	}
+	role, err := i.checkAndCreateRole()
```
Comment (snasovich) [inline pkg/impersonation/impersonation.go:38]: It's weird how this breaks in-between error handling of errors during checking for SA and returning SA if it already exists. Could we move check/creation of role before we check for SA for more logical flow? Maybe there is a reason it has to be this way, but I'm failing to see it since `sa` is not used here.

Comment (cmurphy) [reply pkg/impersonation/impersonation.go]: I refactored this to use the role as the existence check so we don't need to have this double resource check.

File: pkg/impersonation/impersonation.go (hunk 3 above)
Comment (cmurphy) [reply pkg/impersonation/impersonation.go]: fixed in the refactor

File: pkg/impersonation/impersonation.go [hunk 5]
```diff
@@ -182,10 +162,46 @@ func (i *Impersonator) createRole() (*rbacv1.ClusterRole, error) {
...
+	if usernames, ok := extras["username"]; ok {
+		rules = append(rules, rbacv1.PolicyRule{
+			Verbs:         []string{"impersonate"},
+			APIGroups:     []string{"authentication.k8s.io"},
+			Resources:     []string{"userextras/username"},
+			ResourceNames: usernames,
+		})
+	}
+	return rules
+}
```
Comment (snasovich) [inline pkg/impersonation/impersonation.go:204]: Nit - add empty line before next func.

**Review by snasovich — DISMISSED** [review]
LGTM (1 super minor nit).

@dramich @ryansann , please review this one with priority today as Colleen is out on Monday so we want this merged today if it's good to go.

**Review by snasovich — APPROVED** [review]

File: pkg/impersonation/impersonation.go [hunk 6]
```diff
@@ -29,34 +30,39 @@ func New(user user.Info, clusterContext *config.UserContext) Impersonator {
 	return Impersonator{user: user, clusterContext: clusterContext}
 }
 
-func (i *Impersonator) SetUpImpersonation() (*corev1.ServiceAccount, error) {
-	sa, err := i.checkServiceAccount()
+func (i *Impersonator) SetUpImpersonation() error {
+	role, err := i.checkAndUpdateRole()
```
Comment (dramich) [inline pkg/impersonation/impersonation.go:34]: Can we create the rules here since we know we need them in either case? That saves us from creating them twice when we have to fall through to create

File: pkg/impersonation/impersonation.go [hunk 7]
```diff
@@ -182,8 +162,45 @@ func (i *Impersonator) createRole() (*rbacv1.ClusterRole, error) {
 			// in case cache isn't synced yet, use raw client
 			return i.clusterContext.RBAC.ClusterRoles("").Get(name, metav1.GetOptions{})
 		}
+		return role, nil
+	}
+	return role, err
+}
+
+func (i *Impersonator) rulesForUser() []rbacv1.PolicyRule {
```
Comment (dramich) [inline pkg/impersonation/impersonation.go:170]: I like this vs the inline we were doing before.

File: pkg/impersonation/impersonation.go (hunk 6 above)
Comment (cmurphy) [reply pkg/impersonation/impersonation.go]: done

**Review by dramich — APPROVED** [review]

**Review by snasovich — APPROVED** [review]
